[LLVM][DebugInfo] Refactor some code for easier sharing.

cmtice opened the pull request:
Refactor the code that calculates the offsets for the various pieces of the DWARF .debug_names index section, to make it easier to share the code with other tools, such as LLD.

cmtice committed:
[LLVM][DebugInfo] Refactor some code for easier sharing.

Refactor the code that calculates the offsets for the various
pieces of the DWARF .debug_names index section, to make it easier to
share the code with other tools, such as LLD.

cmtice committed:
[LLVM][DebugInfo] Refactor some code for easier sharing.

Clang format fixes.

adrian-prantl commented on llvm/include/llvm/DebugInfo/DWARF/DWARFAcceleratorTable.h:
```diff
@@ -562,6 +562,17 @@ class DWARFDebugNames : public DWARFAcceleratorTable {
   };
 
+  // Offsets for the start of various important tables from the start of the
+  // section.
```
Can you use Doxygen cmments `///`?

cmtice replied:
Done.

adrian-prantl commented on llvm/include/llvm/DebugInfo/DWARF/DWARFAcceleratorTable.h:
```diff
@@ -793,6 +799,14 @@ class DWARFDebugNames : public DWARFAcceleratorTable {
 
+// Calculates the starting offsets for various sections within the
+// debug_names section. This has been pulled out of DWARFDebugNames, to
+// allow easier sharing of the code with other tools, such as LLD.
```
///

felipepiovezan replied:
> This has been pulled out of DWARFDebugNames, to
 // allow easier sharing of the code with other tools, such as LLD.

This is more of a review comment than documentation, I suggest removing this

cmtice replied:
Done.

dwblaikie commented on llvm/lib/DebugInfo/DWARF/DWARFAcceleratorTable.cpp:
```diff
@@ -552,32 +552,52 @@ DWARFDebugNames::NameIndex::extractAbbrev(uint64_t *Offset) {
-  Offset += Hdr.NameCount * SectionOffsetSize;
-  EntryOffsetsBase = Offset;
-  Offset += Hdr.NameCount * SectionOffsetSize;
+  Offset = FindDebugNamesOffsets(Offsets, Offset, Hdr.Format, Hdr);
```
This is strange that we assign to `Offset`, but then overwrite it on the next line?

So what's the return value for? Perhaps if it's needed by the future callers but not right now - maybe it should be in the `Offsets` structure, and that `Offsets` structure should be returned by this function, rather than as an input parameter?

felipepiovezan replied:
Yeah, something feels off here. Suggestion:

1. rename `Offset` to `EndOfHeaderOffset`
2. remove the return value from `FindDebugNamesOffset`
3. declare `Offset` in the line below

cmtice replied:
Did something very similar to what felipe recommended.

cmtice committed:
[LLVM][DebugInfo] Refactor some code for easier sharing.

Change 'FindDebugNamesOffsets' function: make the return type 'void' (return
nothing); update the name to start with lowercase letter.  Break 'offset'
variable in NameIndex::extract into two parts: hdrSize (to be passed to
Header::extract and findDebugNamesOffsets) and 'Offset', to be used for
offsets after call to findDebugNamesOffsets.

Change comments in header file to doxygen comments.

cmtice commented:
I think I have addressed all the review comments; if there's something else I need to do please let me know.

cmtice committed:
[LLVM][DebugInfo] Refactor some code for easier sharing.

Clang format fixes.

felipepiovezan reviewed (approved):
LGTM!

dwblaikie commented on llvm/lib/DebugInfo/DWARF/DWARFAcceleratorTable.cpp:
```diff
@@ -552,32 +552,50 @@ DWARFDebugNames::NameIndex::extractAbbrev(uint64_t *Offset) {
+void llvm::findDebugNamesOffsets(
+    DWARFDebugNames::DWARFDebugNamesOffsets &Offsets, uint64_t HdrSize,
+    dwarf::DwarfFormat Format, const DWARFDebugNames::Header &Hdr) {
+  uint32_t DwarfSize = (Format == llvm::dwarf::DwarfFormat::DWARF64) ? 8 : 4;
```
This should use `dwarf::getDwarfOffsetByteSize`

dwblaikie commented on llvm/lib/DebugInfo/DWARF/DWARFAcceleratorTable.cpp:
```diff
@@ -552,32 +552,50 @@ DWARFDebugNames::NameIndex::extractAbbrev(uint64_t *Offset) {
 }
 
+void llvm::findDebugNamesOffsets(
+    DWARFDebugNames::DWARFDebugNamesOffsets &Offsets, uint64_t HdrSize,
```
`HdrSize` isn't an accurate name for this parameter, I think @felipepiovezan mentioned `EndOfHeaderOffset` as a name in https://github.com/llvm/llvm-project/pull/82153#discussion_r1495104676 which seems accurate.

(the issue here is that while `HdrSize` and `EndOfHeaderOffset` would be identical for the first debug names table in a `.debug_names` section - they aren't identical for the second/other subsequent contributions - `EndOfHeaderOffset` is the offset in the .debug_names section after the header we care about here, which might be the Nth header, not necessarily the first)

dwblaikie commented on llvm/lib/DebugInfo/DWARF/DWARFAcceleratorTable.cpp:
```diff
@@ -552,32 +552,50 @@ DWARFDebugNames::NameIndex::extractAbbrev(uint64_t *Offset) {
 
+void llvm::findDebugNamesOffsets(
+    DWARFDebugNames::DWARFDebugNamesOffsets &Offsets, uint64_t HdrSize,
+    dwarf::DwarfFormat Format, const DWARFDebugNames::Header &Hdr) {
```
Probably remove the `Format` parameter, since it's accessible from the header?

dwblaikie commented on llvm/lib/DebugInfo/DWARF/DWARFAcceleratorTable.cpp:
```diff
@@ -552,32 +552,50 @@ DWARFDebugNames::NameIndex::extractAbbrev(uint64_t *Offset) {
   return Abbrev(Code, dwarf::Tag(Tag), AbbrevOffset, std::move(*AttrEncOr));
 }
 
+void llvm::findDebugNamesOffsets(
```
Probably return the offsets from this function, rather than taking them as an out parameter? It's then clearer that they aren't in/out parameters, or something else more interesting.

dwblaikie reviewed (commented):
A few minor things to address here in follow-up commits, please. Excuse the delay.
